Race condition in JavaParserTypeSolver

`JavaParserTypeSolver` has an `JavaParser` instance that is used to parse the files. At the moment it is not thread safe which makes it throw weird exceptions when parsing from different threads. The limitation here is `JavaParser` because it can only parse one instance at a given time. 

I think a simple fix for this issue, could be make the parser synchronized to avoid the concurrency.

## Stack trace
```
Caused by: java.lang.AssertionError: Invalid token kind 2147483647
	at com.github.javaparser.TokenTypes.getCategory(TokenTypes.java:257)
	at com.github.javaparser.TokenTypes.isComment(TokenTypes.java:63)
	at com.github.javaparser.GeneratedJavaParserTokenManager.CommonTokenAction(GeneratedJavaParserTokenManager.java:106)
	at com.github.javaparser.GeneratedJavaParserTokenManager.getNextToken(GeneratedJavaParserTokenManager.java:2951)
	at com.github.javaparser.GeneratedJavaParser.jj_scan_token(GeneratedJavaParser.java:12682)
	at com.github.javaparser.GeneratedJavaParser.jj_3_1(GeneratedJavaParser.java:8747)
	at com.github.javaparser.GeneratedJavaParser.jj_2_1(GeneratedJavaParser.java:7873)
	at com.github.javaparser.GeneratedJavaParser.CompilationUnit(GeneratedJavaParser.java:74)
	at com.github.javaparser.JavaParser.parse(JavaParser.java:123)
	at com.github.javaparser.symbolsolver.resolution.typesolvers.JavaParserTypeSolver.lambda$parse$1(JavaParserTypeSolver.java:153)
	at com.google.common.cache.LocalCache$LocalManualCache$1.load(LocalCache.java:4853)
	at com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3529)
	at com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2278)
	at com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2155)
	at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2045)
	... 32 more
```

## Unit test
```java
@RepeatedTest(25)
void testTryToSolveTypeWithMultipleThreads() {
    class StressRunnable implements Runnable {

        private final String typeToSolve;
        private final JavaParserTypeSolver javaParserTypeSolver;

        StressRunnable(String typeToSolve, JavaParserTypeSolver javaParserTypeSolver) {
            this.typeToSolve = typeToSolve;
            this.javaParserTypeSolver = javaParserTypeSolver;
        }

        @Override
        public void run() {
            javaParserTypeSolver.tryToSolveType(typeToSolve);
        }
    }

    JavaParserTypeSolver javaParserTypeSolver = JAVA_PARSER_PROVIDER.get();

    CompletableFuture<Void> tasks = CompletableFuture.allOf(
            CompletableFuture.runAsync(new StressRunnable("com.github.javaparser.ast.body.Object",
                    javaParserTypeSolver)),
            CompletableFuture.runAsync(new StressRunnable("com.github.javaparser.ast.comments.Object",
                    javaParserTypeSolver)),
            CompletableFuture.runAsync(new StressRunnable("com.github.javaparser.ast.expr.Object",
                    javaParserTypeSolver)),
            CompletableFuture.runAsync(new StressRunnable("com.github.javaparser.ast.stmt.Object",
                    javaParserTypeSolver))
    );
    assertDoesNotThrow(tasks::join,
            "JavaParserTypeSolve should work properly when called from multiple threads.");
}
```

--- a/javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/resolution/typesolvers/JavaParserTypeSolver.java
+++ b/javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/resolution/typesolvers/JavaParserTypeSolver.java
@@ -21,17 +21,9 @@
 
 package com.github.javaparser.symbolsolver.resolution.typesolvers;
 
-import com.github.javaparser.JavaParser;
-import com.github.javaparser.ParserConfiguration;
-import com.github.javaparser.ast.CompilationUnit;
-import com.github.javaparser.resolution.declarations.ResolvedReferenceTypeDeclaration;
-import com.github.javaparser.symbolsolver.javaparser.Navigator;
-import com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade;
-import com.github.javaparser.symbolsolver.model.resolution.SymbolReference;
-import com.github.javaparser.symbolsolver.model.resolution.TypeSolver;
-import com.github.javaparser.symbolsolver.utils.FileUtils;
-import com.google.common.cache.Cache;
-import com.google.common.cache.CacheBuilder;
+import static com.github.javaparser.ParseStart.COMPILATION_UNIT;
+import static com.github.javaparser.ParserConfiguration.LanguageLevel.BLEEDING_EDGE;
+import static com.github.javaparser.Providers.provider;
 
 import java.io.File;
 import java.io.FileNotFoundException;
@@ -45,9 +37,17 @@ import java.util.Objects;
 import java.util.Optional;
 import java.util.concurrent.ExecutionException;
 
-import static com.github.javaparser.ParseStart.COMPILATION_UNIT;
-import static com.github.javaparser.ParserConfiguration.LanguageLevel.BLEEDING_EDGE;
-import static com.github.javaparser.Providers.provider;
+import com.github.javaparser.JavaParser;
+import com.github.javaparser.ParserConfiguration;
+import com.github.javaparser.ast.CompilationUnit;
+import com.github.javaparser.resolution.declarations.ResolvedReferenceTypeDeclaration;
+import com.github.javaparser.symbolsolver.javaparser.Navigator;
+import com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade;
+import com.github.javaparser.symbolsolver.model.resolution.SymbolReference;
+import com.github.javaparser.symbolsolver.model.resolution.TypeSolver;
+import com.github.javaparser.symbolsolver.utils.FileUtils;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
 
 /**
  * Defines a directory containing source code that should be used for solving symbols.

--- a/javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/resolution/typesolvers/JavaParserTypeSolver.java
+++ b/javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/resolution/typesolvers/JavaParserTypeSolver.java
@@ -21,9 +21,17 @@
 
 package com.github.javaparser.symbolsolver.resolution.typesolvers;
 
-import static com.github.javaparser.ParseStart.COMPILATION_UNIT;
-import static com.github.javaparser.ParserConfiguration.LanguageLevel.BLEEDING_EDGE;
-import static com.github.javaparser.Providers.provider;
+import com.github.javaparser.JavaParser;
+import com.github.javaparser.ParserConfiguration;
+import com.github.javaparser.ast.CompilationUnit;
+import com.github.javaparser.resolution.declarations.ResolvedReferenceTypeDeclaration;
+import com.github.javaparser.symbolsolver.javaparser.Navigator;
+import com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade;
+import com.github.javaparser.symbolsolver.model.resolution.SymbolReference;
+import com.github.javaparser.symbolsolver.model.resolution.TypeSolver;
+import com.github.javaparser.symbolsolver.utils.FileUtils;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
 
 import java.io.File;
 import java.io.FileNotFoundException;
@@ -37,17 +45,9 @@ import java.util.Objects;
 import java.util.Optional;
 import java.util.concurrent.ExecutionException;
 
-import com.github.javaparser.JavaParser;
-import com.github.javaparser.ParserConfiguration;
-import com.github.javaparser.ast.CompilationUnit;
-import com.github.javaparser.resolution.declarations.ResolvedReferenceTypeDeclaration;
-import com.github.javaparser.symbolsolver.javaparser.Navigator;
-import com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade;
-import com.github.javaparser.symbolsolver.model.resolution.SymbolReference;
-import com.github.javaparser.symbolsolver.model.resolution.TypeSolver;
-import com.github.javaparser.symbolsolver.utils.FileUtils;
-import com.google.common.cache.Cache;
-import com.google.common.cache.CacheBuilder;
+import static com.github.javaparser.ParseStart.COMPILATION_UNIT;
+import static com.github.javaparser.ParserConfiguration.LanguageLevel.BLEEDING_EDGE;
+import static com.github.javaparser.Providers.provider;
 
 /**
  * Defines a directory containing source code that should be used for solving symbols.
@@ -149,9 +149,13 @@ public class JavaParserTypeSolver implements TypeSolver {
                     if (!Files.exists(srcFile) || !Files.isRegularFile(srcFile)) {
                         return Optional.empty();
                     }
-                    return javaParser.parse(COMPILATION_UNIT, provider(srcFile))
-                            .getResult()
-                            .map(cu -> cu.setStorage(srcFile));
+
+                    // JavaParser only allow one parse at time.
+                    synchronized (javaParser) {
+                        return javaParser.parse(COMPILATION_UNIT, provider(srcFile))
+                                .getResult()
+                                .map(cu -> cu.setStorage(srcFile));
+                    }
                 } catch (FileNotFoundException e) {
                     throw new RuntimeException("Issue while parsing while type solving: " + srcFile.toAbsolutePath(), e);
                 }

--- a/javaparser-symbol-solver-testing/src/test/java/com/github/javaparser/symbolsolver/resolution/typesolvers/JavaParserTypeSolverTest.java
+++ b/javaparser-symbol-solver-testing/src/test/java/com/github/javaparser/symbolsolver/resolution/typesolvers/JavaParserTypeSolverTest.java
@@ -28,12 +28,14 @@ import com.github.javaparser.symbolsolver.model.resolution.SymbolReference;
 import com.github.javaparser.symbolsolver.utils.LeanParserConfiguration;
 import com.github.javaparser.utils.CodeGenerationUtils;
 import org.junit.jupiter.api.Disabled;
+import org.junit.jupiter.api.RepeatedTest;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.io.TempDir;
 
 import java.io.File;
 import java.io.IOException;
 import java.nio.file.Path;
+import java.util.concurrent.CompletableFuture;
 import java.util.function.Supplier;
 
 import static org.junit.jupiter.api.Assertions.*;
@@ -145,6 +147,45 @@ class JavaParserTypeSolverTest extends AbstractTypeSolverTest<JavaParserTypeSolv
         SymbolReference<ResolvedReferenceTypeDeclaration> x = typeSolver.tryToSolveType("A<>");
 
         assertFalse(x.isSolved());
+    }
+
+    /**
+     * {@link com.github.javaparser.JavaParser} doesn't work across multiple threads.
+     *
+     * This test makes sure the concurrency is handled.
+     */
+    @RepeatedTest(25)
+    void testTryToSolveTypeWithMultipleThreads() {
+        class StressRunnable implements Runnable {
+
+            private final String typeToSolve;
+            private final JavaParserTypeSolver javaParserTypeSolver;
+
+            StressRunnable(String typeToSolve, JavaParserTypeSolver javaParserTypeSolver) {
+                this.typeToSolve = typeToSolve;
+                this.javaParserTypeSolver = javaParserTypeSolver;
+            }
+
+            @Override
+            public void run() {
+                javaParserTypeSolver.tryToSolveType(typeToSolve);
+            }
+        }
+
+        JavaParserTypeSolver javaParserTypeSolver = JAVA_PARSER_PROVIDER.get();
+
+        CompletableFuture<Void> tasks = CompletableFuture.allOf(
+                CompletableFuture.runAsync(new StressRunnable("com.github.javaparser.ast.body.Object",
+                        javaParserTypeSolver)),
+                CompletableFuture.runAsync(new StressRunnable("com.github.javaparser.ast.comments.Object",
+                        javaParserTypeSolver)),
+                CompletableFuture.runAsync(new StressRunnable("com.github.javaparser.ast.expr.Object",
+                        javaParserTypeSolver)),
+                CompletableFuture.runAsync(new StressRunnable("com.github.javaparser.ast.stmt.Object",
+                        javaParserTypeSolver))
+        );
+        assertDoesNotThrow(tasks::join,
+                "JavaParserTypeSolve should work properly when called from multiple threads.");
     }
 
 }

